contracts/implementations/streamr/VehicleStream.sol

```diff
@@ -0,0 +1,102 @@
+
+import "../../shared/Errors.sol" as Errors;
+
+import "@solidstate/contracts/access/access_control/AccessControlInternal.sol";
+import "@openzeppelin/contracts/utils/Strings.sol";
+
+/// TODO Documentation
+contract VehicleStream is AccessControlInternal {
+    string constant DIMO_STREAM_ENS = "streams.dimo.eth";
+    string constant DIMO_STREAM_ENS_VEHICLE = "streams.dimo.eth/vehicle/";
+
+    // TODO Documentation
+    function createStream(uint256 vehicleId) external {
+        address vehicleIdProxyAddress = VehicleStorage
+            .getStorage()
+            .idProxyAddress;
+        IStreamRegistry streamRegistry = IStreamRegistry(
+            StreamrManagerStorage.getStorage().streamrRegistry
+        );
+
+        try INFT(vehicleIdProxyAddress).ownerOf(vehicleId) returns (
+            address vehicleOwner
+        ) {
+            if (vehicleOwner != msg.sender) {
+                revert Errors.Unauthorized(msg.sender);
+            }
+        } catch {
+            revert Errors.InvalidNode(vehicleIdProxyAddress, vehicleId);
+        }
+
+        string memory streamPath = string(
+            abi.encodePacked("/vehicle/", Strings.toString(vehicleId))
+        );
+        streamRegistry.createStreamWithENS(DIMO_STREAM_ENS, streamPath, "{}");
+    }
+
+    // TODO Documentation
+    function subscribe(
+        uint256 vehicleId,
+        address subscriber,
+        uint256 expirationTimestamp
+    ) public {
+        StreamrManagerStorage.Storage storage ss = StreamrManagerStorage
+            .getStorage();
+        VehicleStreamStorage.Storage storage vss = VehicleStreamStorage
+            .getStorage(vehicleId);
+        IStreamRegistry streamRegistry = IStreamRegistry(ss.streamrRegistry);
+
+        string memory streamId = string(
+            abi.encodePacked(
+                DIMO_STREAM_ENS_VEHICLE,
+                Strings.toString(vehicleId)
+            )
+        );
+
+        streamRegistry.grantPermission(
+            streamId,
+            subscriber,
+            IStreamRegistry.PermissionType.Subscribe
+        );
+        streamRegistry.setExpirationTime(
+            streamId,
+            subscriber,
+            IStreamRegistry.PermissionType.Subscribe,
+            expirationTimestamp
+        );
+
+        vss.subscribers.push(subscriber);
+    }
+
+    // TODO Documentation
+    function onTransfer(
+        address /* from */,
+        address /* to */,
+        uint256 vehicleId
+    ) public {
+        StreamrManagerStorage.Storage storage ss = StreamrManagerStorage
+            .getStorage();
+        VehicleStreamStorage.Storage storage vss = VehicleStreamStorage
+            .getStorage(vehicleId);
+        IStreamRegistry streamRegistry = IStreamRegistry(ss.streamrRegistry);
+
+        string memory streamId = vss.streamId;
+
+        for (uint i = 0; i < vss.subscribers.length; i++) {
+            address subscriber = vss.subscribers[i];
```

Comment:
Oh, you guys went with an array. Would love to know the reasoning—it seemed like we were headed towards the delete/recreate thing.

Reply:
Mustn't we also delete this array?

Reply:
Just an experiment as well. Got many reasons against after we did it. I will go with the delete/recreate